components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Cdm/src/Services/CdmMessageProcessor.cs

```diff
@@ -160,47 +159,82 @@ public class CdmMessageProcessor : ICdmClient {
                     await OpenAsync();
                 }
 
-                foreach (var samplesList in _samplesCache.Values) {
-                    if (samplesList.Count == 0 || samplesList[0] == null) {
+                foreach (var record in _samplesCache) {
+                    if (record.Value.Count == 0 || record.Value[0] == null) {
                         _logger.Error("Samples list is empty ...");
                         continue;
                     }
-                    if (!GetEntityData(samplesList[0], out var partitionLocation, out var partitionDelimitor)) {
-                        if(!CreateEntityData(samplesList[0], out partitionLocation, out partitionDelimitor)) {
-                            _logger.Error("Failed to create CDM Entity for {endpointId}/{nodeId}).",
-                                samplesList[0]?.EndpointId, samplesList[0]?.NodeId);
+                    var retry = false;
+                    do {
+                        var partition = GetOrCreateEntityDataPartition(record.Key,
+                            record.Value[0], out var persist, retry);
+                        if (partition == null) {
+                            _logger.Error("Failed to create CDM Entity for " +
+                                "{PublilsherId}/{DataSetWriterId}/{NodeId}).",
+                                record.Value[0].PublisherId, record.Value[0].DataSetWriterId,
+                                record.Value[0].NodeId);
                             continue;
                         }
-                        performSave = true;
-                    }
-                    await _storage.WriteInCsvPartition<MonitoredItemSampleModel>(
-                        partitionLocation, samplesList, partitionDelimitor);
+                        performSave |= persist;
+
+                        var csvTrait = partition.ExhibitsTraits.Item("is.partition.format.CSV");
+                        var partitionUrl = _cdmCorpus.Storage.CorpusPathToAdapterPath(partition.Location);
+                        var partitionDelimitor = csvTrait?.Arguments?.FetchValue("delimiter") ?? kCsvPartitionsDelimiter;
+                        var result = await _storage.WriteInCsvPartition<MonitoredItemSampleModel>(
+                            partitionUrl, record.Value, partitionDelimitor);
+                        if (result == false && retry == false) {
```

Comment:
>if (result == false && retry == false) { [](start = 24, length = 40)

retry = !result && !retry, 
so there is one retry, is this ok, no error handling etc?

Reply:
yes, 1 retry - that will trigger an attempt to create a new partition. In case of error, there is not much we can do about it